File: myExpenses/src/main/java/org/totschnig/myexpenses/ui/TimeButton.java
package org.totschnig.myexpenses.ui;

import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.Context;
import android.util.AttributeSet;

import org.threeten.bp.LocalTime;
import org.threeten.bp.format.DateTimeFormatter;
import org.threeten.bp.format.FormatStyle;

import androidx.annotation.NonNull;
import icepick.State;

import static org.threeten.bp.temporal.ChronoField.HOUR_OF_DAY;
import static org.threeten.bp.temporal.ChronoField.MINUTE_OF_HOUR;

public class TimeButton extends ButtonWithDialog {
  @State
  LocalTime time;

  DateTimeFormatter timeFormatter = DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT);

  public TimeButton(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  @Override
  protected void update() {
    requireTime();
    setText(time.format(timeFormatter));
  }

  @NonNull
  public LocalTime getTime() {
    requireTime();
    return time;
  }

  public void setTime(LocalTime time) {
    this.time = time;
    update();
  }

  @Override
  public Dialog onCreateDialog() {
    requireTime();
    TimePickerDialog.OnTimeSetListener timeSetListener =
        (view, hourOfDay, minute) -> {
          if (time.get(HOUR_OF_DAY) != hourOfDay ||
              time.get(MINUTE_OF_HOUR) != minute) {
            setTime(LocalTime.of(hourOfDay, minute));
            getHost().onValueSet(this);
          }
        };
    return new TimePickerDialog(getContext(),
        timeSetListener,
        time.getHour(),
        time.getMinute(),
        android.text.format.DateFormat.is24HourFormat(getContext())
    );
  }

  private void requireTime() {
    if (time == null) {
      time = LocalTime.now();
    }
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/ui/TimeButton.kt
package org.totschnig.myexpenses.ui

import android.content.Context
import android.text.format.DateFormat.is24HourFormat
import android.util.AttributeSet
import com.evernote.android.state.State
import com.google.android.material.timepicker.MaterialTimePicker
import com.google.android.material.timepicker.TimeFormat
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.preference.PrefKey
import org.totschnig.myexpenses.util.ui.preferredTimePickerBuilder
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

class TimeButton @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : ButtonWithDialog<MaterialTimePicker>(context, attrs, defStyleAttr) {

    @JvmField
    @State
    var time: LocalTime

    init {
        time = LocalTime.now()
    }

    private var timeFormatter: DateTimeFormatter =
        DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT)

    override fun buildDialog() = preferredTimePickerBuilder(context)
        .setHour(time.hour)
        .setMinute(time.minute)
        .build()

    override fun update() {
        text = time.format(timeFormatter)
    }

    fun setTime(time: LocalTime) {
        this.time = time
        update()
    }

    override val fragmentTag: String
        get() = "date_button"

    override fun attachListener(dialogFragment: MaterialTimePicker) {
        dialogFragment.addOnPositiveButtonClickListener {
            context.injector.prefHandler().putInt(PrefKey.TIME_PICKER_INPUT_MODE, dialogFragment.inputMode)
            setTime(LocalTime.of(dialogFragment.hour, dialogFragment.minute))
        }
        dialogFragment.addOnDismissListener {
            dialogShown = false
        }
    }
}
